### PR TITLE
Enable filtering for tasks and custom rewards

### PR DESCRIPTION
**Pull request by JiriChara**
Add search input box next to the tags. When user changes the value of
the input box then it's propagated to the user service as filterQuery
string. filterQuery string is then used in the tasks template to filter
tasks by text.

### CODE REVIEW

**Comment by Alys** [comment]
Fast work! And you included a test! We like that. :)

Could you please add a screenshot of the tasks page with the search input box in it? It doesn't have to be the whole page, just enough to get a good feel for how the box will look with the other elements around it.

Also, a useful tip for github pull requests: If you enter:
 `fixes https://github.com/HabitRPG/habitrpg/issues/5385`
(i.e., the URL for an issue), then github creates a link between the two and it also automatically closes the issue when the PR is merged. Like so:  fixes https://github.com/HabitRPG/habitrpg/issues/5385 :)

**Comment by JiriChara** [comment]
Hi Alys,

thank you for the advice! See the screenshots below:

2015-06-11-100344_469x450_scrot

mouseover:
2015-06-11-100327_739x441_scrot

focus:
2015-06-11-100312_702x374_scrot

no focus/no mouseover
2015-06-11-100209_620x219_scrot

**Comment by Alys** [comment]
@lemoness @AnitaYCheng what do you think of the UI for this?

**Comment by JiriChara** [comment]
I will do little bit of cosmetics, because the search field is not wrapped very nicely on small screens. I also don't like the placeholder text ("Search"). Do you have any idea on how to rename it?

File: website/views/shared/tasks/task.jade [hunk 1]
```diff
@@ -1,4 +1,4 @@
...
+li(bindonce='list', id='task-{{::task.id}}', ng-repeat='task in obj[list.type+"s"] | filter:{text: obj.filterQuery} | conditionalOrderBy: list.view=="dated":"date"', class='task {{Shared.taskClasses(task, user.filters, user.preferences.dayStart, user.lastCron, list.showCompleted, main)}}', ng-click='spell && (list.type != "reward") && castEnd(task, "task", $event)', ng-class='{"cast-target":spell && (list.type != "reward"), "locked-task":obj._locked === true}', popover-trigger='mouseenter', data-popover-html="{{task.popoverOpen ? '' : task.notes | markdown}}", popover-placement="top", popover-append-to-body='{{::modal ? "false":"true"}}', ng-show='shouldShow(task, list, user.preferences)')
```
Comment (crookedneighbor) [inline website/views/shared/tasks/task.jade]: Is there a reason not to filter by notes as well?

File: website/public/js/controllers/filtersCtrl.js [hunk 2]
```diff
@@ -30,6 +31,12 @@ habitrpg.controller("FiltersCtrl", ['$scope', '$rootScope', 'User', 'Shared',
       // User.save();
     };
 
+    $scope.$watch('filterQuery', function (newValue, oldValue) {
```
Comment (crookedneighbor) [inline website/public/js/controllers/filtersCtrl.js]: Do we really need a watch to accomplish this?

Comment (JiriChara) [inline website/public/js/controllers/filtersCtrl.js]: We can assign user to the scope of the controller and then use it's `filterQuery` attribute as a `ng-model` for the search field. I thought that a watcher might be more transparent for the programmers - so they can directly see the relation between user service and filter controller w/o digging into the view and it is easier to test too. I mean, I can still change it if you like!

File: website/views/shared/tasks/task.jade (hunk 1 above)
Comment (JiriChara) [inline website/views/shared/tasks/task.jade]: That's a good point! I will fix it as soon I'll have time.

File: website/public/js/controllers/filtersCtrl.js (hunk 2 above)
Comment (crookedneighbor) [inline website/public/js/controllers/filtersCtrl.js]: Not sure. Still thinking about it. 

In general, we try to limit the number of watches going on. We already have so many :disappointed:

File: website/views/main/filters.jade [hunk 3]
```diff
@@ -1,6 +1,7 @@
 .container-fluid
   .row
     .filters(ng-controller='FiltersCtrl')
+      input.filters-search(type='text', placeholder=env.t('search'), ng-model='filterQuery')
```
Comment (crookedneighbor) [inline website/views/main/filters.jade]: How would this look using bootstrap's input styling instead?

**Comment by JiriChara** [comment]
The new bootstrap design:

2015-06-11-204512_471x98_scrot

**Comment by lemoness** [comment]
Bootstrap looks better, but it seems a little tall.

I'm also not sold on the placement - it seems to imply to me that you use
it to search for tags.Is there a way to have it on the right side of the
page instead?

**Comment by JiriChara** [comment]
The height of the search box is same as the height of those tags. I will put the search field to the right

**Comment by JiriChara** [comment]
@lemoness what do you think about the new position and styling?

2015-06-11-221756_1057x174_scrot

**Comment by lemoness** [comment]
Looks great to me!

**Comment by Alys** [comment]
I like that a lot! It looks very clean and neat.

However what does it look like when there's a lot of tags? I've seen people with three rows of tags. Do they wrap nicely around the search box with a bit of horizontal space to separate them?

**Comment by JiriChara** [comment]
2015-06-11-223230_923x140_scrot

2015-06-11-223201_402x230_scrot

**Comment by Alys** [comment]
I like that. In the first shot, the blank space between the tag controls and the search box could be considered by some to be a waste of space, but I think it looks nice, and uncluttered. The performance on a small screen is great (thanks for thinking to include that).

**Comment by crookedneighbor** [comment]
This will count toward a contributor tier. What's your UUID? It can be found in Settings > API > User Id

**Comment by JiriChara** [comment]
@crookedneighbor cool thanks, my UID is: a85f4815-af74-482e-9d26-5fd50fb071c9 :+1:

**Comment by Alys** [comment]
I've just discovered a problem with this, and unfortunately the revert option no longer works. It's not super critical though, and just for the record, @JiriChara I LOVE this feature and you still get contributor credit for it. :)

If you have a search query in place and you go to another page in the website, then back to the tasks page, the search query is still in place but the search box is empty (as if there is no current search). This will undoubtably lead to people reporting "bugs" that their tasks have disappeared or been deleted.

The way to recover from it is to type anything into the search box and then delete that text, but I think we need a better fix, and also an obvious indication that a search filter has been applied. 

We should probably also clear the search filter when moving away from (or back to) the tasks page. Even if we do that, I think it would help if there was also an indicator that a filter was in place. Perhaps text just above the task columns that says "**showing only tasks containing [search term goes here]**". There could be a "clear filter" button next to that text. (We could also make that text describe any tags that are selected, which would help to reduce the bug reports from tasks being hidden by forgotten tag filters.)

Any other ideas?

Are we planning to do a deployment to production in the next day or two? If so, do we want to revert this manually before then, or do we think we can put in any fix or workaround soon enough that it won't matter?

**Comment by crookedneighbor** [comment]
Ohh, good catch. I know how to fix it.

**Comment by crookedneighbor** [comment]
@Alys this should fix it: https://github.com/HabitRPG/habitrpg/commit/b052f803e4acb079bfdbefa9cc865af8b0d3f06a

**Comment by crookedneighbor** [comment]
@JiriChara I've awarded you your first contributor tier! You can [read more about contributor awards here](http://habitrpg.wikia.com/wiki/Contributor_Rewards).

**Pull request merged**